Issue261/lsh test

mentekid opened the pull request:
Add five test cases to fix #261. Check that the LSH implementation:

1. Increases recall when number of hash tables increases
2. Increases recall when primary hash width increases
3. Decreases recall when number of projections per table increases
4. Very "expensive" parameter sets yield recall higher than 50%
5. Very "cheap" parameter sets yield recall lower than 25%

Test (1) seemed to have a 0.4% false negative, so there's a fail-safe of repeating the test (currently 5 times) and requiring that at least one of the repetitions passes.

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -9,103 +9,211 @@
 
-  arma::Mat<size_t> neighbors;
-  arma::mat distances;
+  math::RandomSeed(time(0));
```
We should remove this for the final commit, so that a test result is reproducible on a single system.

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -9,103 +9,211 @@
 
+double compute_recall(
+    arma::Mat<size_t> LSHneighbors, 
+    arma::Mat<size_t> groundTruth)
```
It's probably better to use const references here: `const arma::Mat<size_t>&` not `arma::Mat<size_t>`.  That will help avoid a copy.

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -9,103 +9,211 @@
+    arma::Mat<size_t> groundTruth)
+{
+  const int n_queries = LSHneighbors.n_cols;
+  const int n_neigh = LSHneighbors.n_rows;
```
Really minor style note: we should avoid underscores in variable names... I'd suggest just `queries` and `neigh`.

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -9,103 +9,211 @@
+  int found_same = 0;
+  for (int q = 0; q < n_queries; ++q)
+  {
+    for (int n = 0; n < n_neigh; ++n)
```
Another minor note: maybe better to use `size_t` here since these will never take negative values?

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -9,103 +9,211 @@
+  //are real neighbors of the query.
+  //LSH's property is that (with high probability), increasing the hash width
+  //will increase recall. Epsilon ensures that if noise lightly affects the 
+  //projections, the test will not fail.
```
We should probably make this a separate `BOOST_AUTO_TEST_CASE()`, even though we'll have to load the data twice.

rcurtin commented:
Looks good to me, my main comments are pretty minor style comments.  I can fix the style after merge and show you what changed if you want, that's a very simple thing for me to do.  Also if you want to add your name to the list of contributors in `src/mlpack/core.hpp` and `COPYRIGHT.txt`, please feel free---this is a pretty significant contribution.

mentekid commented:
I think I can do these changes quickly, no reason to bother you. I'll push them once they're done.

I'll add my name as well, thanks :)

It seems that the checks failed, is that something I should look into? It's weird, everything seems to be working for me...

rcurtin commented:
It looks like a random test failure in the GMM tests, so maybe I need to loosen the tolerance even further.  Let me know when you've made the changes and I'll go ahead and merge. :)

mentekid commented:
I made the style changes.
About the separate test case, I am not sure which test you mean, could you clarify? I'll leave home for a while and do that once I get back

rcurtin commented:
Ah, yeah, what I mean is that we should split each of the five things you are testing into separate test cases.  Like this:

```
//Test: Run LSH with varying number of tables, keeping all other parameters 
//constant. Compute the recall, i.e. the number of reported neighbors that
//are real neighbors of the query.
//LSH's property is that (with high probability), increasing the number of
//tables will increase recall. Epsilon ensures that if noise lightly affects
//the projections, the test will not fail.
//This produces false negatives, so we attempt the test numTries times and
//only declare failure if all of them fail.
BOOST_AUTO_TEST_CASE(RecallTest)
{
  // code here
}

//Test: Run LSH with varying hash width, keeping all other parameters 
//constant. Compute the recall, i.e. the number of reported neighbors that
//are real neighbors of the query.
//LSH's property is that (with high probability), increasing the hash width
//will increase recall. Epsilon ensures that if noise lightly affects the 
//projections, the test will not fail.
BOOST_AUTO_TEST_CASE(HashWidthTest)
{
  // code here
}
```

I hope I've described that well enough, let me know if not.  The basic idea is that we can split each of the tests so that a user can run only one test at a time.  The disadvantage to this approach is that you probably have to load a dataset multiple times, but that's not so huge of a deal since the dataset will be relatively small.

mentekid commented:
Oh I see. Yeah that would be better. I'll do that tomorrow. I think the
style issues are solved at my latest commit, please tell me if there's
anything that I forgot :)

On Tue, Apr 5, 2016 at 10:33 PM, Ryan Curtin <notifications@github.com>
wrote:

> Ah, yeah, what I mean is that we should split each of the five things you
> are testing into separate test cases. Like this:
>
> //Test: Run LSH with varying number of tables, keeping all other parameters
> //constant. Compute the recall, i.e. the number of reported neighbors that
> //are real neighbors of the query.
> //LSH's property is that (with high probability), increasing the number of
> //tables will increase recall. Epsilon ensures that if noise lightly affects
> //the projections, the test will not fail.
> //This produces false negatives, so we attempt the test numTries times and
> //only declare failure if all of them fail.
> BOOST_AUTO_TEST_CASE(RecallTest)
> {
>   // code here
> }
>
> //Test: Run LSH with varying hash width, keeping all other parameters
> //constant. Compute the recall, i.e. the number of reported neighbors that
> //are real neighbors of the query.
> //LSH's property is that (with high probability), increasing the hash width
> //will increase recall. Epsilon ensures that if noise lightly affects the
> //projections, the test will not fail.
> BOOST_AUTO_TEST_CASE(HashWidthTest)
> {
>   // code here
> }
>
> I hope I've described that well enough, let me know if not. The basic idea
> is that we can split each of the tests so that a user can run only one test
> at a time. The disadvantage to this approach is that you probably have to
> load a dataset multiple times, but that's not so huge of a deal since the
> dataset will be relatively small.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/mlpack/mlpack/pull/605#issuecomment-205992606>
>

rcurtin commented:
There are still some style issues, but don't worry about it too much, I can easily fix when I merge it.  Basically the things I see are variable names with underscores (we prefer `variableName` over `variable_name`), horizontal whitespace (we prefer `a + b` vs. `a+b`), and maybe some places where opening brackets should be on newlines.  Like I said it's pretty trivial stuff, but I value consistency pretty highly across the codebase. :)

mentekid commented:
I split the tests into different boost testcases and fixed whitespaces underscores and brackets wherever I could find them. I also commented out the random seed as you said.

rcurtin commented:
I made some style fixes in bc6b9e7 and wrote `ComputeRecall()` as a nicer expression in ba826b1.  Let me know if my changes broke anything (but the tests still seem to work).
